File: leetcode/1187.make-array-strictly-increasing/solution.cpp
#include "../../common.hpp"

size_t max_len1 = 0;
size_t max_len2 = 0;

// https://leetcode.com/problems/make-array-strictly-increasing
class Solution {
public:
  static int makeArrayIncreasing(vector<int> &arr1, vector<int> &arr2) {
    std::sort(arr2.begin(), arr2.end());
    vector<std::pair<int, int>> prev{{0, -1}};

    for (int now_val : arr1) {
      vector<std::pair<int, int>> curr;
      for (auto [last_mod, last_val] : prev) {
        // auto [last_mod, last_val] = last_mod_val;
        if (last_val < now_val) {
          if (!curr.empty() && curr.back().first == last_mod)
            curr.back().second = std::min(curr.back().second, now_val);
          else
            curr.emplace_back(last_mod, now_val);
        }
        if (auto it = std::upper_bound(arr2.begin(), arr2.end(), last_val);
            it != arr2.end())
          curr.emplace_back(last_mod + 1, *it);
      }
      max_len1 += curr.size();
      prev = std::move(curr);
    }
    return prev.empty() ? -1 : prev.front().first;
  }
};

// https://leetcode.com/problems/make-array-strictly-increasing
class Solution2 {
public:
  static int makeArrayIncreasing(vector<int> &arr1, vector<int> &arr2) {
    std::sort(arr2.begin(), arr2.end());
    vector<std::pair<int, int>> prev{{0, -1}};

    for (int now_val : arr1) {
      vector<std::pair<int, int>> curr;
      for (auto [last_mod, last_val] : prev) {
        // auto [last_mod, last_val] = last_mod_val;
        if (last_val < now_val) {
          if (!curr.empty() && curr.back().first == last_mod)
            curr.back().second = std::min(curr.back().second, now_val);
          else
            curr.emplace_back(last_mod, now_val);
        }
        if (auto it = std::upper_bound(arr2.begin(), arr2.end(), last_val);
            it != arr2.end() && (curr.empty() || curr.back().second > *it))
          curr.emplace_back(last_mod + 1, *it);
      }
      // DEBUG(curr);
      max_len2 += curr.size();
      prev = std::move(curr);
    }
    return prev.empty() ? -1 : prev.front().first;
  }
};

int main() {
  srand(0);
  // vector<int> v1{1, 5, 3, 6, 7}, v2{1, 3, 2, 4};
  // vector<int> v1{1, 5, 3, 6, 7}, v2{1, 3, 4};
  // vector<int> v1{1, 5, 3, 6, 7}, v2{1, 6, 3, 3};
  vector<int> v1, v2;

  for (int t = 0; t < 100; ++t) {
    for (int i = 0; i < 5000; ++i) {
      v1.push_back(rand() % 10000);
    }
    for (int i = 0; i < 50; ++i) {
      v2.push_back(rand() % 10000);
    }
    auto n1 = Solution::makeArrayIncreasing(v1, v2);
    auto n2 = Solution2::makeArrayIncreasing(v1, v2);
    assert(n1 == n2);
  }
  DEBUG(max_len1, max_len2);
}

File: leetcode/1187.make-array-strictly-increasing/solution.rs
fn main() {
    let v1 = vec![1, 5, 3, 6, 7];

    // let v2 = vec![1, 3, 2, 4];
    // let v2 = vec![4,3,1];
    let v2 = vec![1, 6, 3, 3];

    dbg!(Solution::make_array_increasing(v1, v2));
}

struct Solution;
// https://leetcode.com/problems/make-array-strictly-increasing
use std::cmp;

impl Solution {
    pub fn make_array_increasing(arr1: Vec<i32>, mut arr2: Vec<i32>) -> i32 {
        arr2.sort();
        let mut prev = vec![(0i32, -1)];
        for now_val in arr1.into_iter() {
            let mut curr = vec![];
            for (last_mod, last_val) in prev.into_iter() {
                if last_val < now_val {
                    match curr.last_mut() {
                        Some((n, v)) if *n == last_mod => {
                            *v = cmp::min(*v, now_val)
                        }
                        _ => curr.push((last_mod, now_val)),
                    }
                }
                let index = arr2.partition_point(|&v| v <= last_val);
                if let Some(&v) = arr2.get(index) {
                    // This line is optional
                    if curr.last().filter(|p| p.1 <= v).is_none() {
                        curr.push((last_mod + 1, v));
                    }
                }
            }
            prev = curr;
        }
        prev.into_iter().next().map_or(-1, |(c, _)| c)
    }
}
